### PR TITLE
`<functional>`: is compatible with `/Zc:alignedNew-` now

### LINKED ISSUE #2711: <functional>: incompatible with `/std:c++latest` and `/Zc:alignedNew-`
**Command-line test case**
```
C:\Temp>type repro.cpp
#include <functional>

int main() {}

C:\Temp>cl /std:c++latest /Zc:alignedNew- /nologo /EHsc /W4 /WX .\repro.cpp

C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1304): error C2760: syntax error: '{' was unexpected here; expected ')'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1304): error C3878: syntax error: unexpected token 'identifier' following 'expression'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1304): note: error recovery skipped: 'identifier'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1304): error C2760: syntax error: '}' was unexpected here; expected ';'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C2760: syntax error: '{' was unexpected here; expected ')'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token 'alignof' following 'expression'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): note: error recovery skipped: 'alignof ( identifier'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C2760: syntax error: ')' was unexpected here; expected ';'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token ')' following 'expression_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token ')' following 'statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token ')' following 'statement_seq'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): note: missing one of:  '}' ?
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C2760: syntax error: ')' was unexpected here; expected '}'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token ')' following 'compound_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): error C3878: syntax error: unexpected token ')' following 'selection_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1314): note: error recovery skipped: ')'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C2760: syntax error: '{' was unexpected here; expected ')'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token 'alignof' following 'expression'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): note: error recovery skipped: 'alignof ( identifier'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C2760: syntax error: ')' was unexpected here; expected ';'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token ')' following 'expression_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token ')' following 'statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token ')' following 'statement_seq'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): note: missing one of:  '}' ?
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C2760: syntax error: ')' was unexpected here; expected '}'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token ')' following 'compound_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): error C3878: syntax error: unexpected token ')' following 'selection_statement'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\include\functional(1335): note: error recovery skipped: ')
```

**Expected behavior**
The program should compile.

**STL version**
    ```
    Microsoft Visual Studio Community 2022 Preview
    Version 17.3.0 Preview 1.0
    ```

**Additional context**
Reference: https://docs.microsoft.com/en-us/cpp/build/reference/zc-alignednew?view=msvc-170
Also tracked by DevCom-10036384 and Microsoft-Internal VSO-1538233 / AB#1538233.

Problem code places:
https://github.com/microsoft/STL/blob/488e7893861b52b0c5bcb7b91061b0db7009ae73/stl/inc/functional#L1302
https://github.com/microsoft/STL/blob/488e7893861b52b0c5bcb7b91061b0db7009ae73/stl/inc/functional#L1312
https://github.com/microsoft/STL/blob/488e7893861b52b0c5bcb7b91061b0db7009ae73/stl/inc/functional#L1333
https://github.com/microsoft/STL/blob/488e7893861b52b0c5bcb7b91061b0db7009ae73/stl/inc/functional#L1343

### PR DESCRIPTION
**Pull request by fsb4000**
fixes #2711 

I also found that:

```c++
#include <iostream>

using namespace std;

int main() {
#ifdef __cpp_aligned_new
    cout << "has aligned new\n";
#else
    cout << "nothing\n";
#endif
}
```
works differently on `cl.exe` and `clang-cl.exe`

```
$ clang-cl main.cpp /std:c++14 /Zc:alignedNew /EHsc
$ main
has aligned new
```
```
$ cl main.cpp /std:c++14 /Zc:alignedNew /EHsc
$ main
nothing
```

this is why I replaced all

```
#ifdef __cpp_aligned_new
```

to
```
#if defined(__cpp_aligned_new) && _HAS_CXX17
```

Should I report about it to LLVM and add "Transition" comment?

### CODE REVIEW

File: stl/inc/functional
Comment (AlexGuteniev) [inline stl/inc/functional:1316]: What C++17? `move_only_function` is under C++23 block.

**Review by AlexGuteniev — COMMENTED** [review]
I'm concerned about changing existing code to rely on C++17 mode

File: tests/std/tests/GH_002711_Zc_alignedNew-/env.lst
Comment (AlexGuteniev) [inline tests/std/tests/GH_002711_Zc_alignedNew-/env.lst:6]: I'm not sure if this is sufficient test. Does `move_only_function` actually work in this mode?

**Comment by fsb4000** [comment]
cl defines `__cpp_aligned_new` for C++17 and up (with `/Zc:alignedNew`)
clang-cl defines `__cpp_aligned_new` for C++14 and up (with `/Zc:alignedNew`)
I added `_HAS_CXX17` for compatibility between `clang-cl` and `cl`
Am I wrong?

**Comment by AlexGuteniev** [comment]
> cl defines `__cpp_aligned_new` for C++17 and up (with `/Zc:alignedNew`)
> clang-cl defines `__cpp_aligned_new` for C++14 and up (with `/Zc:alignedNew`)
> I added `_HAS_CXX17` for compatibility between `clang-cl` and `cl`
> Am I wrong?

That's an interesting question then. Does clang-cl compile with `/std:c++14` and `/Zc:alignedNew` before this change?

If no, the change is good.

If yes, then we are in pretty bad situation with no good options.

**Review by AlexGuteniev — APPROVED** [review]
Ok then I agree

File: tests/std/tests/GH_002711_Zc_alignedNew-/env.lst
Comment (AlexGuteniev) [reply tests/std/tests/GH_002711_Zc_alignedNew-/env.lst:6]: Yes please

**Comment by CaseyCarter** [comment]
> works differently on `cl.exe` and `clang-cl.exe`

This is a separate issue, which should be handled in a separate PR (if at all). Feel free to report upstream to the clang-cl devs that cl ignores `/Zc:alignedNew` in pre-C++17 modes and that they could do so as well if they like.

**Comment by fsb4000** [comment]
> This is a separate issue, which should be handled in a separate PR (if at all)

@CaseyCarter
current tests will not pass if I revert `#if defined(__cpp_aligned_new) && _HAS_CXX17` to `#ifdef __cpp_aligned_new`
```
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-cl.EXE" "C:\Dev\STL\tests\std\tests\GH_000690_overaligned_function\test.cpp" "-IC:\Dev\STL\out\build\x64\out\inc" "-IC:\Dev\STL\llvm-project\libcxx\test\support" "-IC:\Dev\STL\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/sdl" "/WX" "/Zc:strictStrings" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "-fno-ms-compatibility" "-fno-delayed-template-parsing" "/EHsc" "/MD" "/std:c++14" "/w14640" "/Zc:threadSafeInit-" "/Zc:alignedNew" "-m64" "-FeC:\Dev\STL\out\build\x64\tests\std\tests\GH_000690_overaligned_function\Output\55\GH_000690_overaligned_function.exe" "-link" "-LIBPATH:C:\Dev\STL\out\build\x64\out\lib\amd64" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31424\lib\x64" "/MANIFEST:EMBED"
Exit Code: 1
Standard Error:
--
In file included from C:\Dev\STL\tests\std\tests\GH_000690_overaligned_function\test.cpp:9:
In file included from C:\Dev\STL\out\build\x64\out\inc\functional:14:
C:\Dev\STL\out\build\x64\out\inc\xmemory(1425,60): error: cannot initialize a value of type 'std::align_val_t' with an rvalue of type 'unsigned long long'
                _Pbuf = ::operator new (_Size, align_val_t{alignof(_Ty)}, nothrow);
                                                           ^~~~~~~~~~~~
C:\Dev\STL\out\build\x64\out\inc\xmemory(1445,47): error: cannot initialize a value of type 'std::align_val_t' with an rvalue of type 'unsigned long long'
        ::operator delete (_Pbuf, align_val_t{alignof(_Ty)});
                                              ^~~~~~~~~~~~
2 errors generated.
```

**Comment by fsb4000** [comment]
I created: LLVM-55389

File: stl/inc/functional
Comment (StephanTLavavej) [inline stl/inc/functional:1305]: These should be `_CSTD` qualified. Example:

https://github.com/microsoft/STL/blob/313aa593223e78b458ac50a5a46357601eebfe00/stl/inc/fstream#L405

File: stl/inc/functional
Comment (StephanTLavavej) [inline stl/inc/functional:1359]: The `::operator new()` calls are throwing, so the `_aligned_malloc()` call needs to detect allocation failure and call `_Xbad_alloc()`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
FYI @strega-nil-ms I pushed minor changes after you approved.

I was initially concerned about mismatch - adding `__cpp_aligned_new` logic to otherwise unguarded code would create an incompatibility between C++14 and C++17 mode (as C++17 implies this by default). However, the behavioral change here is specific to C++23 mode. Therefore, only mixing `/std:c++latest` and `/std:c++latest /Zc:alignedNew-` would be a problem, and this is already true for our other `__cpp_aligned_new` guards.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: stl/inc/functional
Comment (CaseyCarter) [inline stl/inc/functional:1305]: I'm concerned that the treatment of overaligned `move_only_function` targets in `/Zc:alignedNew-` mode here is inconsistent with how we treat overaligned types everywhere else in the STL. Admittedly we couldn't change the handling for things that existed in C++14 without breaking existing code and `move_only_function` is new, but I'm concerned that `/Zc:alignedNew-` users will _want_ the C++14 behavior and will report this inconsistency as a bug.

Am I being overly paranoid? Am I just wrong? Is there some urgent reason to handle `alignedNew-` differently here?

Comment (frederick-vs-ja) [reply stl/inc/functional:1305]: I think the existing handling for `/Zc:alignedNew-` is broken, as it may allocate improperly aligned storage. I believe that creating an object at an improperly aligned address is an unreasonable request, because it seems that the Standard doesn't support it ([[basic.life]/(1.1)](https://eel.is/c++draft/basic.life#1.1))... although it's already non-Standard to use `/Zc:alignedNew-` in C++23 mode.

If user-replaced (normal) `operator new` is wanted, I think we should allocate more bytes (perhaps `sizeof(_Vt) + alignof(_Vt)`?), manually align the address, and then record the offset (or something equivalent) before the object.

Comment (AlexGuteniev) [reply stl/inc/functional:1305]: See https://developercommunity.visualstudio.com/t/10036384#T-N10037977

> The only reason I am using `/Zc:alignedNew-` at all is to eliminate the space/time overhead of overaligned allocations due to the inability to customize `__STDCPP_DEFAULT_NEW_ALIGNMENT__` while using a global operator new that provides stronger alignment guarantees.

So the user wants to specifically avoid library-provided overalignment, and they are doing overalignment for entire global new. By doing anything except global new, we defeat user's purpose.

**Comment by Nick-Kooij** [comment]
Hello,
Thanks all for working on this issue.  As the developer who raise the issue in the first place, I'd just like to clarify that my purpose in using `/Zc:alignedNew-` is really a workaround to eliminate the overhead of calling overaligned new/delete.

The ideal solution for me would be to add a compiler switch `/Zc:defaultNewAlignment={std::align_val_t}` where the `__STDCPP_DEFAULT_NEW_ALIGNMENT__` compile-time alignment threshold where C++17 overaligned new/delete are called:
- `void* operator new ( std::size_t count, std::align_val_t al );`
- `void operator delete ( void* ptr, std::align_val_t al ) noexcept;`
- And similar array/`noexcept` overloads--anything taking an additional `std::align_val_t` parameter.

In other words, implement a switch symmetric to Clang's [-fnew-alignment](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fnew-alignment).  See [Add compiler switch to change __STDCPP_DEFAULT_NEW_ALIGNMENT__](https://developercommunity.visualstudio.com/t/Add-compiler-switch-to-change-__STDCPP_D/373283).

---

Some background:

1.  I have overloaded global operator new/delete providing stronger alignment guarantees than the default `__STDCPP_DEFAULT_NEW_ALIGNMENT__`.  Specifically alignment beyond a single 64 byte cache line is aligned to 128 bytes (two cache lines).  (This meshes well with x64 prefetching logic and avoids false sharing.)

2.  By default with C++17 compilation (without `/Zc:alignedNew`) `new/delete` expressions for types with alignment beyond `__STDCPP_DEFAULT_NEW_ALIGNMENT__` will call the C++17 overaligned operator new/delete overloads:
- `void* operator new ( std::size_t count, std::align_val_t al );`
- `void operator delete ( void* ptr, std::align_val_t al ) noexcept;`
- ...

This is a fantastic new C++17 feature allowing over-aligned types to be dynamically allocated.  *Ideally I would like this feature enabled, rather than disabling it using `/Zc:alignedNew-`.*

Consider allocation of a 256 byte type, with 128 byte alignment, by default without `/Zc:alignedNew-`.  As `__STDCPP_DEFAULT_NEW_ALIGNMENT__` can't be we have **we now have runtime overhead which I wish to avoid:**
- bloated `operator new` and `operator delete` call sites
- runtime overhead of calling an overaligned operator new

**The goal is to eliminate this runtime overhead.**

While the runtime overhead within the overloaded overaligned new/delete can be somewhat mitigated using `[[likely]]/[[unlikely]]` on the alignment check, nevertheless there is a runtime overhead due to unnecessary branches.


```
<public: static void __cdecl TaskSpawn<struct ChainTask>::spawn<int>(int &&)>:
               	push	rbx
               	sub	rsp, 32
               	mov	rbx, rcx
               	mov	edx, 128 // additional std::align_val_t parameter
               	mov	ecx, 256 // object size
               	call	 ??2@YAPEAX_KW4align_val_t@std@@@Z // overaligned operator new call
                ...
```

This bloats each over-aligned `operator new/delete` call site, and unnecessarily calls an overload that contains additional alignment logic.

Ideally we want this:

```
<public: static void __cdecl TaskSpawn<struct ChainTask>::spawn<int>(int &&)>:
               	push	rbx
               	sub	rsp, 32
               	mov	rbx, rcx
               	mov	ecx, 256 // object size
               	call	 ??2@YAPEAX_K@Z // regular operator new called
                ...
```

Best Regards,
Nick

**Comment by StephanTLavavej** [comment]
We talked about this at the weekly maintainer meeting - we believe that this PR should be changed to follow the same strategy as the rest of the STL currently uses for `/Zc:alignedNew-`, which is to directly call `operator new` and not attempt to add additional alignment. As @frederick-vs-ja mentioned, this will potentially not handle overaligned types correctly (in the case where `operator new` hasn't been replaced to provide such alignment), but this is consistent with existing practice, and attempting to add additional alignment would interfere with scenarios like @Nick-Kooij's. We conclude that if users want the STL to automatically handle overaligned types correctly, they should be using the C++17 Standard setting - by setting `/Zc:alignedNew-` they are responsible for their own destiny.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm speculatively mirroring this to the MSVC-internal repo - further changes can be pushed, but please notify me.

File: stl/inc/functional
Comment (CaseyCarter) [inline stl/inc/functional:1346]: Ditto:
```suggestion
#ifdef __cpp_aligned_new
                if constexpr (alignof(_Vt) > __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
                   ::operator delete (_Ptr, align_val_t{alignof(_Vt)});
                } else
#endif // __cpp_aligned_new
                {
                   ::operator delete(_Ptr);
                }
```

File: stl/inc/functional
Comment (CaseyCarter) [inline stl/inc/functional:1360]: Ditto:
```suggestion
#ifdef __cpp_aligned_new
    if constexpr (alignof(_Vt) > __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
        _Ptr = ::operator new (sizeof(_Vt), align_val_t{alignof(_Vt)});
    } else
#endif // __cpp_aligned_new
    {
        _Ptr = ::operator new(sizeof(_Vt));
    }
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Review by CaseyCarter — APPROVED** [review]
Applied my suggested changes after testing locally.

File: stl/inc/functional
Comment (CaseyCarter) [inline stl/inc/functional:1302]: Oops. Credit to @StephanTLavavej
```suggestion
    _STL_INTERNAL_STATIC_ASSERT(_Align > __STDCPP_DEFAULT_NEW_ALIGNMENT__);
```

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for *aligning* these *new* features with our existing tolerance of this option! :joy_cat: :rocket: :tada:

**Pull request merged**
